File: scripts/project_profile.py
#!/usr/bin/env python3
from pathlib import Path
import json
import os

MANAGED_MARKER = "<!-- WEB-AGENT-KIT:MANAGED -->"
PROJECT_CONTEXT_START = "<!-- WEB-AGENT-KIT:PROJECT-CONTEXT:START -->"
PROJECT_CONTEXT_END = "<!-- WEB-AGENT-KIT:PROJECT-CONTEXT:END -->"

IGNORE_DIRS = {
    ".git", ".agents", ".agent-core", "node_modules", "vendor", "bin", "obj",
    ".next", "dist", "build", "coverage", ".idea", ".vscode", ".cache"
}

ALLOWED_HIDDEN_DIRS = {".github", ".devcontainer"}

IMPORTANT_ROOT_FILES = {
    "package.json", "composer.json", "tsconfig.json", "pyproject.toml", "go.mod",
    "Cargo.toml", "Dockerfile", "docker-compose.yml", "docker-compose.yaml",
    "compose.yml", "compose.yaml", "pnpm-workspace.yaml", "turbo.json", "nx.json",
    "vite.config.ts", "vite.config.js", "next.config.js", "next.config.mjs",
    "next.config.ts", "artisan"
}

CATEGORY_LABELS = {
    "languages": "Languages",
    "frontend": "Frontend",
    "backend": "Backend",
    "api-realtime": "API / Realtime",
    "auth": "Authentication / Authorization",
    "data": "Data / ORM",
    "distributed-jobs": "Distributed / Jobs",
    "storage-media": "Storage / Media",
    "testing": "Testing",
    "observability-performance": "Observability / Performance",
    "devops-delivery": "DevOps / Delivery",
    "packages": "Package Management",
    "integrations": "Integrations",
    "architecture-patterns": "Architecture"
}


def load_json(path):
    try:
        return json.loads(Path(path).read_text(encoding="utf-8"))
    except Exception:
        return {}


def detect_project_name(project):
    project = Path(project).resolve()

    package = load_json(project / "package.json")
    name = package.get("name")
    if isinstance(name, str) and name.strip():
        return name.strip()

    composer = load_json(project / "composer.json")
    name = composer.get("name")
    if isinstance(name, str) and name.strip():
        return name.strip()

    sln = sorted(project.glob("*.sln"))
    if sln:
        return sln[0].stem

    csproj = sorted(project.glob("*.csproj"))
    if csproj:
        return csproj[0].stem

    return project.name


def build_structure_summary(project, max_depth=2, max_entries=70):
    """Build a shallow structural map without walking dependency/build directories."""
    project = Path(project).resolve()
    directories = set()
    root_files = set()

    for current_root, dirnames, filenames in os.walk(project):
        current = Path(current_root)
        try:
            rel_root = current.relative_to(project)
        except ValueError:
            continue

        # Prune ignored and irrelevant hidden directories before os.walk descends into them.
        dirnames[:] = [
            name for name in dirnames
            if name not in IGNORE_DIRS
            and (not name.startswith(".") or name in ALLOWED_HIDDEN_DIRS)
        ]

        depth = 0 if rel_root == Path(".") else len(rel_root.parts)
        if depth >= max_depth:
            dirnames[:] = []

        for dirname in dirnames:
            rel_dir = (rel_root / dirname) if rel_root != Path(".") else Path(dirname)
            if 1 <= len(rel_dir.parts) <= max_depth:
                directories.add(rel_dir.as_posix() + "/")

        if depth == 0:
            for filename in filenames:
                p = Path(filename)
                if filename in IMPORTANT_ROOT_FILES or p.suffix.lower() in {".sln", ".csproj"}:
                    root_files.add(filename)

        if len(directories) + len(root_files) >= max_entries * 2:
            # Enough candidates for a deterministic capped summary.
            break

    entries = sorted(directories) + sorted(root_files)
    return entries[:max_entries]


def group_detected_skills(skills, kit_root):
    catalog = load_json(Path(kit_root) / "pack" / "router" / "skill-catalog.json")
    category_by_skill = {
        item.get("name"): item.get("category")
        for item in catalog.get("skills", [])
        if isinstance(item, dict)
    }

    grouped = {}
    for skill in skills:
        category = category_by_skill.get(skill)
        if category not in CATEGORY_LABELS:
            continue
        grouped.setdefault(CATEGORY_LABELS[category], []).append(skill)

    return {
        label: sorted(set(values))
        for label, values in grouped.items()
        if values
    }


def build_project_profile(project, kit_root, version):
    project = Path(project).resolve()
    cfg = load_json(project / ".agent-kit.json")
    index = load_json(project / ".agent-core" / "index" / "project-index.json")
    skills = cfg.get("detected_skills", []) if isinstance(cfg.get("detected_skills", []), list) else []

    return {
        "name": detect_project_name(project),
        "directory": project.name,
        "domain": "web",
        "kit_version": version,
        "technology_groups": group_detected_skills(skills, kit_root),
        "detected_skills": sorted(set(skills)),
        "structure": build_structure_summary(project),
        "manifests": index.get("manifests", []),
        "config_files": index.get("config_files", []),
        "test_roots": index.get("test_roots", []),
        "migration_roots": index.get("migration_roots", []),
        "source": "Generated from lightweight structural discovery; source contents are not embedded."
    }


def _bullet_list(values, empty="None detected"):
    if not values:
        return f"- {empty}"
    return "\n".join(f"- `{value}`" for value in values)


def render_project_context(profile):
    technology_lines = []
    for label, values in profile.get("technology_groups", {}).items():
        technology_lines.append(f"- **{label}:** " + ", ".join(f"`{v}`" for v in values))
    if not technology_lines:
        technology_lines.append("- No framework/language-specific skills were strongly detected yet; use routed discovery before assuming a stack.")

    structure = profile.get("structure", [])
    structure_block = "\n".join(structure) if structure else "(No shallow structure detected)"

    return f"""{MANAGED_MARKER}
{PROJECT_CONTEXT_START}
# Project Agent Context — {profile.get('name', 'Web Project')}

This section is generated by the Web Development Agent Kit installer and is intentionally lightweight.
Use it for **initial routing**, then verify exact behavior with targeted discovery before making changes.

## Project Identity

- **Project name:** `{profile.get('name', 'unknown')}`
- **Project directory:** `{profile.get('directory', 'unknown')}`
- **Development domain:** Web Development
- **Web Kit version:** `{profile.get('kit_version', 'unknown')}`

## Detected Technology Stack

{chr(10).join(technology_lines)}

## Project Structure — Shallow Routing Map

```text
{structure_block}
```

This is a routing map, not permission to read every listed directory.
The Context Router still controls which files/symbols each agent may inspect.

## Important Manifests / Build Files

{_bullet_list(profile.get('manifests', []))}

## Important Configuration Files

{_bullet_list(profile.get('config_files', []))}

## Test Roots

{_bullet_list(profile.get('test_roots', []))}

## Migration / Data Change Roots

{_bullet_list(profile.get('migration_roots', []))}

## Project-aware Routing Rules

- Read this generated project context before selecting agents or skills.
- Treat detected technologies as routing evidence, not as proof that every part of the repository uses them.
- Start from the user-requested feature and route only the relevant project slice.
- Do not read the entire repository because it appears in the structure map.
- Prefer existing project conventions and ownership discovered from the code over generic framework assumptions.
- If the repository structure or stack changes materially, regenerate the project profile with the Web Kit update/install command.
- The machine-readable copy of this profile is `.agent-core/index/project-profile.json`.

{PROJECT_CONTEXT_END}

---
"""


def is_kit_managed(content):
    if MANAGED_MARKER in content:
        return True
    # Recognize pre-project-profile Web Kit AGENTS files so upgrades can migrate them safely.
    return "# Web Development Agent Kit v1" in content and "## Canonical workflow authority" in content


def apply_project_profile(project, kit_root=None, version=None, force_agents=False):
    project = Path(project).resolve()
    kit_root = Path(kit_root).resolve() if kit_root else Path(__file__).resolve().parents[1]
    if version is None:
        version_file = kit_root / "VERSION"
        version = version_file.read_text(encoding="utf-8").strip() if version_file.exists() else "unknown"

    profile = build_project_profile(project, kit_root, version)
    template = (kit_root / "AGENTS.template.md").read_text(encoding="utf-8")
    generated = render_project_context(profile) + template.rstrip() + "\n"

    agents_path = project / "AGENTS.md"
    fallback_path = project / "AGENTS.web-kit.md"

    if agents_path.exists():
        existing = agents_path.read_text(encoding="utf-8", errors="ignore")
        if force_agents or is_kit_managed(existing):
            agents_path.write_text(generated, encoding="utf-8")
            if fallback_path.exists():
                fallback_path.unlink()
            output_path = agents_path
            action = "updated project-aware AGENTS.md"
        else:
            fallback_path.write_text(generated, encoding="utf-8")
            output_path = fallback_path
            action = "preserved existing AGENTS.md; wrote project-aware AGENTS.web-kit.md"
    else:
        agents_path.write_text(generated, encoding="utf-8")
        if fallback_path.exists():
            fallback_path.unlink()
        output_path = agents_path
        action = "wrote project-aware AGENTS.md"

    index_dir = project / ".agent-core" / "index"
    index_dir.mkdir(parents=True, exist_ok=True)
    (index_dir / "project-profile.json").write_text(json.dumps(profile, indent=2), encoding="utf-8")

    cfg_path = project / ".agent-kit.json"
    cfg = load_json(cfg_path)
    if cfg:
        cfg["project"] = {
            "name": profile["name"],
            "directory": profile["directory"],
            "domain": profile["domain"],
            "technology_groups": profile["technology_groups"]
        }
        cfg_path.write_text(json.dumps(cfg, indent=2), encoding="utf-8")

    return {
        "path": str(output_path),
        "action": action,
        "profile": profile
    }


def main():
    import argparse

    parser = argparse.ArgumentParser(description="Generate project-aware AGENTS.md context")
    parser.add_argument("project", nargs="?", default=".")
    parser.add_argument("--force-agents", action="store_true")
    args = parser.parse_args()

    result = apply_project_profile(args.project, force_agents=args.force_agents)
    print(result["action"])
    print(f"Project: {result['profile']['name']}")


if __name__ == "__main__":
    main()

File: scripts/project-profile.mjs
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const IGNORE_DIRS = new Set([
  ".git", ".agents", ".agent-core", "node_modules", "vendor", "bin", "obj",
  ".next", "dist", "build", "coverage", ".idea", ".vscode", ".cache", "graphify-out",
]);
const ALLOWED_HIDDEN_DIRS = new Set([".github", ".devcontainer"]);
const IMPORTANT_ROOT_FILES = new Set([
  "package.json", "composer.json", "tsconfig.json", "pyproject.toml", "go.mod",
  "Cargo.toml", "Dockerfile", "docker-compose.yml", "docker-compose.yaml",
  "compose.yml", "compose.yaml", "pnpm-workspace.yaml", "turbo.json", "nx.json",
  "vite.config.ts", "vite.config.js", "next.config.js", "next.config.mjs",
  "next.config.ts", "artisan",
]);
const GRAPHIFY_MCP_CONFIGS = [
  ".mcp.json", "mcp.json", ".cursor/mcp.json", ".gemini/settings.json",
  ".vscode/mcp.json", ".codex/config.toml",
];
const CATEGORY_LABELS = {
  languages: "Languages",
  frontend: "Frontend",
  backend: "Backend",
  "api-realtime": "API / Realtime",
  auth: "Authentication / Authorization",
  data: "Data / ORM",
  "distributed-jobs": "Distributed / Jobs",
  "storage-media": "Storage / Media",
  testing: "Testing",
  "observability-performance": "Observability / Performance",
  "devops-delivery": "DevOps / Delivery",
  packages: "Package Management",
  integrations: "Integrations",
  "architecture-patterns": "Architecture",
};

export function readJson(file) {
  try {
    return JSON.parse(fs.readFileSync(file, "utf8"));
  } catch {
    return {};
  }
}

export function writeJson(file, value) {
  fs.mkdirSync(path.dirname(file), { recursive: true });
  fs.writeFileSync(file, `${JSON.stringify(value, null, 2)}\n`, "utf8");
}

export function detectProjectName(project) {
  const pkg = readJson(path.join(project, "package.json"));
  if (typeof pkg.name === "string" && pkg.name.trim()) return pkg.name.trim();

  const composer = readJson(path.join(project, "composer.json"));
  if (typeof composer.name === "string" && composer.name.trim()) return composer.name.trim();

  try {
    const entries = fs.readdirSync(project, { withFileTypes: true });
    const sln = entries.find((e) => e.isFile() && e.name.endsWith(".sln"));
    if (sln) return path.parse(sln.name).name;
    const csproj = entries.find((e) => e.isFile() && e.name.endsWith(".csproj"));
    if (csproj) return path.parse(csproj.name).name;
  } catch {}

  return path.basename(project);
}

function looksLikeGraphifyCodeMcp(text) {
  const lower = String(text || "").toLowerCase();
  return lower.includes("graphify.serve") || lower.includes("graphify-mcp") ||
    (lower.includes("graphify") && lower.includes("graphify-out/graph.json"));
}

export function detectGraphify(project) {
  const graphPath = path.join(project, "graphify-out", "graph.json");
  const graphReady = fs.existsSync(graphPath) && fs.statSync(graphPath).isFile();
  const configHits = [];

  for (const rel of GRAPHIFY_MCP_CONFIGS) {
    const file = path.join(project, rel);
    if (!fs.existsSync(file)) continue;
    try {
      if (looksLikeGraphifyCodeMcp(fs.readFileSync(file, "utf8"))) configHits.push(rel);
    } catch {}
  }

  const installState = readJson(path.join(project, ".agent-core", "state", "graphify-install.json"));
  const detected = graphReady || configHits.length > 0 || installState.configured === true;
  let note;
  if (graphReady) {
    note = "Graphify graph snapshot detected. Use graph-assisted navigation only after the managed freshness gate reports a clean graph; verify behavior from exact source.";
  } else if (detected) {
    note = "Graphify integration detected but no project graph snapshot is ready. Use standard routing until graphify-out/graph.json exists.";
  } else {
    note = "Graphify not detected in this project; use standard repository routing.";
  }

  return {
    detected,
    graph_ready: graphReady,
    graph_path: graphReady ? "graphify-out/graph.json" : null,
    local_mcp_configured: configHits.length > 0,
    config_files: configHits.sort(),
    routing_mode: graphReady ? "graphify-assisted" : "standard",
    fallback_mode: "standard",
    runtime_tool_required: graphReady,
    source_authority: "repository source",
    note,
  };
}

export function buildStructureSummary(project, maxDepth = 2, maxEntries = 70) {
  const directories = new Set();
  const rootFiles = new Set();

  function walk(current, depth) {
    if (directories.size + rootFiles.size >= maxEntries * 2) return;
    let entries;
    try {
      entries = fs.readdirSync(current, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (IGNORE_DIRS.has(entry.name)) continue;
        if (entry.name.startsWith(".") && !ALLOWED_HIDDEN_DIRS.has(entry.name)) continue;
        const full = path.join(current, entry.name);
        const rel = path.relative(project, full).split(path.sep).join("/");
        const relDepth = rel.split("/").filter(Boolean).length;
        if (relDepth >= 1 && relDepth <= maxDepth) directories.add(`${rel}/`);
        if (depth + 1 < maxDepth) walk(full, depth + 1);
      } else if (depth === 0) {
        if (IMPORTANT_ROOT_FILES.has(entry.name) || entry.name.endsWith(".sln") || entry.name.endsWith(".csproj")) {
          rootFiles.add(entry.name);
        }
      }
    }
  }

  walk(project, 0);
  return [...directories, ...rootFiles].sort().slice(0, maxEntries);
}

export function groupDetectedSkills(skills, kitRoot) {
  const catalog = readJson(path.join(kitRoot, "pack", "router", "skill-catalog.json"));
  const categoryBySkill = new Map();
  for (const item of Array.isArray(catalog.skills) ? catalog.skills : []) {
    if (item && typeof item === "object") categoryBySkill.set(item.name, item.category);
  }

  const grouped = {};
  for (const skill of skills) {
    const category = categoryBySkill.get(skill);
    const label = CATEGORY_LABELS[category];
    if (!label) continue;
    if (!grouped[label]) grouped[label] = [];
    grouped[label].push(skill);
  }

  for (const label of Object.keys(grouped)) {
    grouped[label] = [...new Set(grouped[label])].sort();
  }
  return grouped;
}

export function buildProjectProfile(project, kitRoot, version) {
  const cfg = readJson(path.join(project, ".agent-kit.json"));
  const index = readJson(path.join(project, ".agent-core", "index", "project-index.json"));
  const skills = Array.isArray(cfg.detected_skills) ? cfg.detected_skills : [];

  return {
    name: detectProjectName(project),
    directory: path.basename(project),
    domain: "web",
    kit_version: version,
    technology_groups: groupDetectedSkills(skills, kitRoot),
    detected_skills: [...new Set(skills)].sort(),
    capabilities: { graphify: detectGraphify(project) },
    structure: buildStructureSummary(project),
    manifests: Array.isArray(index.manifests) ? index.manifests : [],
    config_files: Array.isArray(index.config_files) ? index.config_files : [],
    test_roots: Array.isArray(index.test_roots) ? index.test_roots : [],
    migration_roots: Array.isArray(index.migration_roots) ? index.migration_roots : [],
    source: "Generated from lightweight structural discovery; source contents are not embedded.",
  };
}

export function applyProjectProfile(project, { kitRoot, version } = {}) {
  project = path.resolve(project);
  kitRoot = kitRoot ? path.resolve(kitRoot) : path.resolve(__dirname, "..");
  if (!version) {
    try { version = fs.readFileSync(path.join(kitRoot, "VERSION"), "utf8").trim(); }
    catch { version = "unknown"; }
  }

  const profile = buildProjectProfile(project, kitRoot, version);
  const profilePath = path.join(project, ".agent-core", "index", "project-profile.json");
  writeJson(profilePath, profile);

  const cfgPath = path.join(project, ".agent-kit.json");
  const cfg = readJson(cfgPath);
  if (Object.keys(cfg).length) {
    cfg.project = {
      name: profile.name,
      directory: profile.directory,
      domain: profile.domain,
      technology_groups: profile.technology_groups,
    };
    cfg.capabilities = profile.capabilities;
    writeJson(cfgPath, cfg);
  }

  return { path: profilePath, action: "updated machine-readable project profile", profile };
}

function isMain() {
  return process.argv[1] && path.resolve(process.argv[1]) === __filename;
}

if (isMain()) {
  const args = process.argv.slice(2);
  const projectArg = args.find((arg) => !arg.startsWith("-")) || ".";
  const result = applyProjectProfile(projectArg);
  console.log(result.action);
  console.log(`Project: ${result.profile.name}`);
}
